crates/bevy_ecs/src/schedule/graph/graph_map.rs

```diff
@@ -0,0 +1,480 @@
+//! `Graph<DIRECTED>` is a graph datastructure where node values are mapping
+//! keys.
+//! Based on the `GraphMap` datastructure from [`petgraph`].
+//!
+//! [`petgraph`]: https://docs.rs/petgraph/0.6.5/petgraph/
+
+use bevy_utils::{hashbrown::HashSet, AHasher};
+use core::{
+    fmt,
+    hash::{BuildHasher, BuildHasherDefault, Hash},
+};
+use indexmap::IndexMap;
+
+use super::{tarjan_scc::TarjanScc, NodeId};
+
+use Direction::{Incoming, Outgoing};
+
+/// A `Graph` with undirected edges.
+///
+/// For example, an edge between *1* and *2* is equivalent to an edge between
+/// *2* and *1*.
+pub type UnGraph<S = BuildHasherDefault<AHasher>> = Graph<false, S>;
+
+/// A `Graph` with directed edges.
+///
+/// For example, an edge from *1* to *2* is distinct from an edge from *2* to
+/// *1*.
+pub type DiGraph<S = BuildHasherDefault<AHasher>> = Graph<true, S>;
+
+/// `Graph<DIRECTED>` is a graph datastructure using an associative array
+/// of its node weights `NodeId`.
+///
+/// It uses an combined adjacency list and sparse adjacency matrix
+/// representation, using **O(|V| + |E|)** space, and allows testing for edge
```

Comment:
```suggestion
/// representation, using **O(|N| + |E|)** space, and allows testing for edge
```
I think this is supposed to be N for nodes instead of vertices? I see that it comes from petgraph, but unsure what we would want here.